serverless/internal/storage/fs/fs.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fs provides a simple filesystem log storage implementation.
+package fs
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/internal/storage/fs/layout"
+)
+
+// Storage is a serverless storage implementation which uses files to store tree state.
+// The on-disk structure is:
+//  <rootDir>/leaves/aa/bb/cc/ddeeff...
+//  <rootDir>/leaves/pending/
+//  <rootDir>/seq/aa/bb/cc/ddeeff...
+//  <rootDir>/tile/<level>/aa/bb/ccddee...
+//  <rootDir>/state
+type Storage struct {
+	rootDir string
+	state   api.LogState
+}
+
+// New returns an FS instance initialised from the filesystem.
+func New(rootDir string) (*Storage, error) {
+	fi, err := os.Stat(rootDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat %q: %w", rootDir, err)
+	}
+
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("%q is not a directory", rootDir)
+	}
+
+	s, err := loadLogState(filepath.Join(rootDir, layout.StatePath))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Storage{
+		rootDir: rootDir,
+		state:   *s,
+	}, nil
+}
+
+// Create creates a new filesystem hierarchy and returns a Storage representation for it.
+func Create(rootDir string, emptyHash []byte) (*Storage, error) {
+	_, err := os.Stat(rootDir)
+	if err == nil {
+		return nil, fmt.Errorf("%q %w", rootDir, os.ErrExist)
+	}
+
+	if err := os.MkdirAll(rootDir, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create directory %q: %w", rootDir, err)
+	}
+
+	for _, sfx := range []string{"leaves/pending", "seq", "tree"} {
+		path := filepath.Join(rootDir, sfx)
+		if err := os.MkdirAll(path, 0755); err != nil {
+			return nil, fmt.Errorf("failed to create directory %q: %w", path, err)
+		}
+	}
+
+	fs := &Storage{
+		rootDir: rootDir,
+	}
+
+	logState := api.LogState{
+		Size:     0,
+		RootHash: emptyHash,
+		Hashes:   [][]byte{},
+	}
+
+	if err := fs.UpdateState(logState); err != nil {
+		return nil, err
+	}
+	return fs, nil
+}
+
+// LogState returns the current LogState.
+func (fs *Storage) LogState() api.LogState {
+	return fs.state
+}
+
+// UpdateState updates the stored log state.
+func (fs *Storage) UpdateState(newState api.LogState) error {
```

Comment:
May want to document the threading model here. "this is not thread safe" is totally legit.

Reply:
Done (on the struct in general above), also spotted and fixed a bad thing here.